WordPress/gutenberg
# Disallow drag from Nav list view into block canvas

## What?


Disables dragging a block from the Navigation block list view into the editor canvas (or any other context).

Closes https://github.com/WordPress/gutenberg/issues/46994

## Why?


As shown in https://github.com/WordPress/gutenberg/issues/46994 it's possible to drag any block in the Nav block list view into the editor canvas and it will be removed from the Nav block and end up in the editor.

This behaviour is unexpected in this context and should be disallowed.

Note: it has been confirmed that this behaviour is _expected_ in the main List View 🤷 


## How?


Allows Draggables to define a list of DropZone `targets` with which they are compatible.

Allows Dropzones to define a name/identifier for themselves.

When an event occurs over a given Dropzone, it will check the given Draggable for targets. If they exist then it will check for its own name within the Draggables `targets`. Only if found will it accept the Draggable. Otherwise the drop will be rejected.

Note that it has been necessary to introduce new state to track the drag targets. Why? Because the browser's drag event's `dataTransfer` is read only during the `dragover`. Apparently this is by design. As a result we need to store informartion about the dragOrigin in state in order that we can read it during the `dragover` event and avoid showing UI affordances indicating a drop is possible (when it isn't).

### ⚠️ Known issues

Currently the drag "chip" (the black icon that follows your cursor as you drag) has a "positive" icon as you drag. It would be ideal if we were able to make that a "X" during the drag if the dropzone is not compatible. However as it's defined on the draggable there's not currently an easy way to do that without adding more state. Therefore this can be tackled in a follow up.

## Testing Instructions





The key here is ensuring:

1. You cannot successfully complete drag and drop out of the Nav block list view into the editor canvas.
2. _All_ existing drag and drop functionality is preserved.

- First try doing some normal drag and drop within canvas.
- Try dragging within and outside of of main (not Nav block!) List View.
- Now try the same but for the Nav block's list view. You should still be able to drag and drop within it but not be able to successfully complete a drag outside.

### Testing Instructions for Keyboard


## Screenshots or screencast 

https://user-images.githubusercontent.com/444434/216307819-477f457a-d87b-4d42-98f1-adba2e233a5c.mp4
